### PR TITLE
Ensemble recentering application

### PR DESCRIPTION
**Pull request by guillaumevernieres**
Why?
-------
Re-centering of the LETKF members is needed for the hybrid EnVAR in implementation at EMC. 

Dependency
---------------
Implementation of the ensemble re-centering app from [oops pr#452](https://github.com/JCSDA/oops/pull/452)
See [oops issue #451](https://github.com/JCSDA/oops/issues/451) for "more" details.

What this app does:
------------------------
- Scratches your back
- Re-centers an ensemble around your favorite state

See example of usage in the associated test. 

Check with gcc and intel and different HPC:
- [x] "TOL" is needed ? 
- [x] After #238 is merged, check if NOTRAPFPE is needed

### CODE REVIEW

File: test/CMakeLists.txt [hunk 1]
```diff
@@ -379,6 +381,12 @@ soca_add_test( NAME parameters_bump_loc
                NOCOMPARE
                TEST_DEPENDS test_soca_enspert )
 
+soca_add_test( NAME ensrecenter
+               EXE  soca_ensrecenter.x
+               TOL  3.0e-4 0
+               NOTRAPFPE
```
Comment (travissluka) [inline test/CMakeLists.txt]: I'll try on discover, but the extra large tolerance and disabling fpe trapping shouldn't be needed

Comment (travissluka) [reply test/CMakeLists.txt]: okay, I was wrong, turns out floating point exceptions do occur, darn ice. Should be addressed separately, eventually. 

Large tolerance should still be removed if possible.

Comment (guillaumevernieres) [reply test/CMakeLists.txt]: @travissluka ... I do have a fix somewhere for sea-ice. I'll have a look tomorrow.

Comment (travissluka) [reply test/CMakeLists.txt]: Is the sea ice issue just that the halo for ice is not initialized or updated?

**Comment by travissluka** [comment]
Ocean variables look good, I checked to make sure that the recentered ensemble mean matches the output from the identity forecast. The mean doesn't match for `aice` though, would that be expected?
Screenshot from 2019-11-05 08-29-54

**Comment by guillaumevernieres** [comment]
> Ocean variables look good, I checked to make sure that the recentered ensemble mean matches the output from the identity forecast. The mean doesn't match for `aice` though, would that be expected?
> Screenshot from 2019-11-05 08-29-54

@travissluka ... man, I thought you didn't care about sea-ice!!!!
Yes, expected, we check for out of bound values for cicen inside of add increment.

**Comment by guillaumevernieres** [comment]
@travissluka , Just tested on Hera. You are correct, the default small tolerances works.

**Review by travissluka — APPROVED** [review]

**Comment by ytremolet** [comment]
Just merged the oops PR for recentering.

**Comment by travissluka** [comment]
@guillaumevernieres I think this is ready to merge then. It seems we can't remove the NOTRAPFPE yet until other FPEs are addressed

**Comment by guillaumevernieres** [comment]
> @guillaumevernieres I think this is ready to merge then. It seems we can't remove the NOTRAPFPE yet until other FPEs are addressed

@travissluka merging #238 first might be a good idea

**Comment by guillaumevernieres** [comment]
> @guillaumevernieres I think this is ready to merge then. It seems we can't remove the NOTRAPFPE yet until other FPEs are addressed

Ha ... I obviously can't/don't read!

**Pull request merged**